File: web/frontend/src/lib/quran-db.ts
import { ratio, fragmentScore } from "./levenshtein";
import type { QuranVerse } from "./types";

const _BSM_PHONEMES_JOINED = "bismi allahi arraHmaani arraHiimi";
const _BSM_PHONEME_TOKENS = "b i s m i | a l l a h i | a r r a H m aa n i | a r r a H ii m i".split(
  " ",
);

export interface QuranTokenEncoder {
  encodeRawPhonemes(rawPhonemes: string): number[];
}

export interface QuranCandidate {
  surah: number;
  ayah: number;
  ayah_end?: number | null;
  text: string;
  phonemes_joined: string;
  phoneme_token_ids: number[];
  stage_a_score: number;
  raw_score: number;
  bonus: number;
  kind: "single" | "span";
  surah_rank?: number;
}

export interface CandidateRetrieval {
  singles: QuranCandidate[];
  spans: QuranCandidate[];
  combined: QuranCandidate[];
}

interface RetrievalOptions {
  maxSpan?: number;
  hint?: [number, number] | null;
  singleLimit?: number;
  topSurahs?: number;
  spanLimit?: number;
}

export function partialRatio(short: string, long: string): number {
  if (!short || !long) return 0.0;
  if (short.length > long.length) [short, long] = [long, short];
  const window = short.length;
  let best = 0.0;
  for (let i = 0; i <= Math.max(0, long.length - window); i++) {
    const r = ratio(short, long.slice(i, i + window));
    if (r > best) {
      best = r;
      if (best === 1.0) break;
    }
  }
  return best;
}

export class QuranDB {
  verses: QuranVerse[];
  private _byRef: Map<string, QuranVerse> = new Map();
  private _bySurah: Map<number, QuranVerse[]> = new Map();

  constructor(
    data: QuranVerse[],
    private tokenEncoder?: QuranTokenEncoder,
  ) {
    this.verses = data;
    for (const v of data) {
      this._byRef.set(`${v.surah}:${v.ayah}`, v);
      const arr = this._bySurah.get(v.surah) ?? [];
      arr.push(v);
      this._bySurah.set(v.surah, arr);

      v.phoneme_tokens = v.phonemes.trim().split(/\s+/).filter(Boolean);

      if (
        v.ayah === 1 &&
        v.surah !== 1 &&
        v.surah !== 9 &&
        v.phonemes_joined.startsWith(_BSM_PHONEMES_JOINED)
      ) {
        const stripped = v.phonemes_joined.slice(_BSM_PHONEMES_JOINED.length).trim();
        v.phonemes_joined_no_bsm = stripped || null;

        let strippedTokens = v.phoneme_tokens.slice(_BSM_PHONEME_TOKENS.length);
        if (strippedTokens[0] === "|") {
          strippedTokens = strippedTokens.slice(1);
        }
        v.phoneme_tokens_no_bsm = strippedTokens.length ? strippedTokens : null;
      } else {
        v.phonemes_joined_no_bsm = null;
        v.phoneme_tokens_no_bsm = null;
      }

      v.phonemes_joined_ns = v.phonemes_joined.replace(/ /g, "");
      v.phonemes_joined_no_bsm_ns = v.phonemes_joined_no_bsm
        ? v.phonemes_joined_no_bsm.replace(/ /g, "")
        : null;

      if (this.tokenEncoder) {
        v.phoneme_token_ids = this.tokenEncoder.encodeRawPhonemes(v.phonemes);
        v.phoneme_token_ids_no_bsm = v.phoneme_tokens_no_bsm
          ? this.tokenEncoder.encodeRawPhonemes(v.phoneme_tokens_no_bsm.join(" "))
          : null;
      } else {
        v.phoneme_token_ids = [];
        v.phoneme_token_ids_no_bsm = null;
      }

      v.word_token_ends = this._computeWordTokenEnds(v.phoneme_tokens);
    }
  }

  get totalVerses(): number {
    return this.verses.length;
  }

  get surahCount(): number {
    return this._bySurah.size;
  }

  getVerse(surah: number, ayah: number): QuranVerse | undefined {
    return this._byRef.get(`${surah}:${ayah}`);
  }

  getSurah(surah: number): QuranVerse[] {
    return this._bySurah.get(surah) ?? [];
  }

  getNextVerse(surah: number, ayah: number): QuranVerse | undefined {
    const verses = this._bySurah.get(surah) ?? [];
    for (let i = 0; i < verses.length; i++) {
      if (verses[i].ayah === ayah) {
        if (i + 1 < verses.length) return verses[i + 1];
        const nextSurah = this._bySurah.get(surah + 1) ?? [];
        return nextSurah[0];
      }
    }
    return undefined;
  }

  /** Return candidates for verses whose non-Bsm phoneme token IDs are short (≤ maxTokens). */
  getShortVerseCandidates(maxTokens = 15): QuranCandidate[] {
    const result: QuranCandidate[] = [];
    for (const v of this.verses) {
      const ids = v.phoneme_token_ids_no_bsm ?? v.phoneme_token_ids ?? [];
      if (ids.length === 0 || ids.length > maxTokens) continue;
      result.push({
        surah: v.surah,
        ayah: v.ayah,
        text: v.phonemes_joined,
        phonemes_joined: v.phonemes_joined,
        phoneme_token_ids: ids,
        stage_a_score: 0,
        raw_score: 0,
        bonus: 0,
        kind: "single",
      });
    }
    return result;
  }

  search(text: string, topK = 5): (QuranVerse & { score: number })[] {
    const scored: (QuranVerse & { score: number })[] = [];
    for (const v of this.verses) {
      const score = ratio(text, v.phonemes_joined);
      scored.push({ ...v, score });
    }
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, topK);
  }

  retrieveCandidates(
    text: string,
    {
      maxSpan = 4,
      hint = null,
      singleLimit = 32,
      topSurahs = 3,
      spanLimit = 32,
    }: RetrievalOptions = {},
  ): CandidateRetrieval {
    if (!text.trim()) {
      return { singles: [], spans: [], combined: [] };
    }

    const bonuses = this._continuationBonuses(hint);
    const textWords = text.split(/\s+/).filter(Boolean);
    const noSpaceText = text.replace(/ /g, "");
    const scored: [QuranVerse, number, number, number][] = [];

    for (const v of this.verses) {
      let raw = ratio(text, v.phonemes_joined);
      const verseWords = v.phoneme_words;
      const sharedWordCount = Math.min(textWords.length, verseWords.length);
      if (sharedWordCount > 0) {
        const textPrefix = textWords.slice(0, sharedWordCount).join(" ");
        const versePrefix = verseWords.slice(0, sharedWordCount).join(" ");
        raw = Math.max(raw, ratio(textPrefix, versePrefix));
      }
      if (noSpaceText.length <= 10) {
        raw = Math.max(raw, this._shortQueryBoost(noSpaceText, v));
      }
      if (v.phonemes_joined_no_bsm) {
        raw = Math.max(raw, ratio(text, v.phonemes_joined_no_bsm));
        if (noSpaceText.length <= 10) {
          raw = Math.max(raw, this._shortQueryBoost(noSpaceText, v, true));
        }
      }
      const bonus = bonuses.get(`${v.surah}:${v.ayah}`) ?? 0.0;
      if (bonus > 0) {
        const sp = QuranDB._suffixPrefixScore(text, v.phonemes_joined);
        raw = Math.max(raw, sp);
      }
      scored.push([v, raw, bonus, Math.min(raw + bonus, 1.0)]);
    }
    scored.sort((a, b) => b[3] - a[3]);

    const pass2Surahs: number[] = [];
    for (let i = 0; i < scored.length && pass2Surahs.length < topSurahs; i++) {
      const surah = scored[i][0].surah;
      if (!pass2Surahs.includes(surah)) {
        pass2Surahs.push(surah);
      }
    }

    if (noSpaceText.length >= 8) {
      let resorted = false;
      for (let i = 0; i < scored.length; i++) {
        const [v, raw, bonus] = scored[i];
        if (noSpaceText.length >= (v.phonemes_joined_ns?.length ?? 0) * 0.8) continue;

        let frag = fragmentScore(noSpaceText, v.phonemes_joined_ns ?? "");
        if (v.phonemes_joined_no_bsm_ns) {
          frag = Math.max(frag, fragmentScore(noSpaceText, v.phonemes_joined_no_bsm_ns));
        }
        if (frag > raw) {
          const boosted = raw + (frag - raw) * 0.7;
          scored[i] = [v, boosted, bonus, Math.min(boosted + bonus, 1.0)];
          resorted = true;
        }
      }
      if (resorted) scored.sort((a, b) => b[3] - a[3]);
    }

    const singles = scored
      .slice(0, singleLimit)
      .map(([v, raw, bonus, total]) =>
        this._candidateFromVerse(v, raw, bonus, total),
      );

    const spans: QuranCandidate[] = [];
    for (let surahRank = 0; surahRank < pass2Surahs.length; surahRank++) {
      const surah = pass2Surahs[surahRank];
      const verses = this._bySurah.get(surah) ?? [];
      for (let i = 0; i < verses.length; i++) {
        for (let span = 2; span <= maxSpan; span++) {
          if (i + span > verses.length) break;
          const chunk = verses.slice(i, i + span);
          const spanText = this._joinedSpanPhonemes(chunk);
          let raw = ratio(text, spanText);
          const spanWords = spanText.split(/\s+/).filter(Boolean);
          const sharedWordCount = Math.min(textWords.length, spanWords.length);
          if (sharedWordCount > 0) {
            const textPrefix = textWords.slice(0, sharedWordCount).join(" ");
            const spanPrefix = spanWords.slice(0, sharedWordCount).join(" ");
            raw = Math.max(raw, ratio(textPrefix, spanPrefix));
          }
          const bonus = bonuses.get(`${chunk[0].surah}:${chunk[0].ayah}`) ?? 0.0;
          const score = Math.min(raw + bonus, 1.0);
          spans.push(this._candidateFromSpan(chunk, raw, bonus, score, surahRank));
        }
      }
    }
    spans.sort((a, b) => b.stage_a_score - a.stage_a_score);

    return {
      singles,
      spans: spans.slice(0, spanLimit),
      combined: singles.concat(spans.slice(0, spanLimit)),
    };
  }

  matchVerse(
    text: string,
    threshold = 0.3,
    maxSpan = 3,
    hint: [number, number] | null = null,
    returnTopK = 0,
  ): Record<string, any> | null {
    const retrieved = this.retrieveCandidates(text, {
      maxSpan,
      hint,
      singleLimit: Math.max(returnTopK, 5),
      topSurahs: 20,
      spanLimit: 64,
    });

    const ranked = retrieved.combined
      .slice()
      .sort((a, b) => b.stage_a_score - a.stage_a_score);
    const best = ranked[0];
    if (!best || best.stage_a_score < threshold) {
      return null;
    }

    const result: Record<string, any> = {
      surah: best.surah,
      ayah: best.ayah,
      ayah_end: best.ayah_end,
      text: best.text,
      phonemes_joined: best.phonemes_joined,
      score: best.stage_a_score,
      raw_score: best.raw_score,
      bonus: best.bonus,
    };
    if (returnTopK > 0) {
      result.runners_up = retrieved.singles.slice(0, returnTopK).map((candidate) => ({
        surah: candidate.surah,
        ayah: candidate.ayah,
        raw_score: Math.round(candidate.raw_score * 1000) / 1000,
        bonus: Math.round(candidate.bonus * 1000) / 1000,
        score: Math.round(candidate.stage_a_score * 1000) / 1000,
        phonemes_joined: candidate.phonemes_joined.slice(0, 60),
      }));
    }
    return result;
  }

  private _computeWordTokenEnds(tokens: readonly string[]): number[] {
    const ends: number[] = [];
    let rawTokenIndex = 0;
    let lastWasBoundary = true;

    for (const token of tokens) {
      rawTokenIndex++;
      if (token === "|") {
        lastWasBoundary = true;
        continue;
      }
      if (lastWasBoundary) {
        ends.push(rawTokenIndex);
      } else {
        ends[ends.length - 1] = rawTokenIndex;
      }
      lastWasBoundary = false;
    }

    return ends;
  }

  private _candidateFromVerse(
    verse: QuranVerse,
    raw: number,
    bonus: number,
    total: number,
  ): QuranCandidate {
    return {
      surah: verse.surah,
      ayah: verse.ayah,
      ayah_end: verse.ayah,
      text: verse.text_uthmani,
      phonemes_joined: verse.phonemes_joined,
      phoneme_token_ids:
        verse.phoneme_token_ids_no_bsm ?? verse.phoneme_token_ids ?? [],
      stage_a_score: total,
      raw_score: raw,
      bonus,
      kind: "single",
    };
  }

  private _candidateFromSpan(
    chunk: QuranVerse[],
    raw: number,
    bonus: number,
    total: number,
    surahRank: number,
  ): QuranCandidate {
    const first = chunk[0];
    const tokenIds: number[] = [];
    const firstIds = first.phoneme_token_ids_no_bsm ?? first.phoneme_token_ids ?? [];
    tokenIds.push(...firstIds);
    for (let i = 1; i < chunk.length; i++) {
      tokenIds.push(...(chunk[i].phoneme_token_ids ?? []));
    }

    return {
      surah: first.surah,
      ayah: first.ayah,
      ayah_end: chunk[chunk.length - 1].ayah,
      text: chunk.map((verse) => verse.text_uthmani).join(" "),
      phonemes_joined: this._joinedSpanPhonemes(chunk),
      phoneme_token_ids: tokenIds,
      stage_a_score: total,
      raw_score: raw,
      bonus,
      kind: "span",
      surah_rank: surahRank,
    };
  }

  private _joinedSpanPhonemes(chunk: QuranVerse[]): string {
    const firstText = chunk[0].phonemes_joined_no_bsm ?? chunk[0].phonemes_joined;
    return [firstText]
      .concat(chunk.slice(1).map((verse) => verse.phonemes_joined))
      .join(" ");
  }

  private _shortQueryBoost(
    noSpaceText: string,
    verse: QuranVerse,
    useNoBsm = false,
  ): number {
    const candidate = useNoBsm
      ? verse.phonemes_joined_no_bsm_ns ?? verse.phonemes_joined_ns ?? ""
      : verse.phonemes_joined_ns ?? "";
    if (!candidate) return 0;

    const prefixWindow = Math.min(candidate.length, noSpaceText.length + 6);
    const prefix = ratio(noSpaceText, candidate.slice(0, prefixWindow));
    const firstWord = useNoBsm
      ? (verse.phonemes_joined_no_bsm ?? "").split(" ")[0] ?? ""
      : verse.phoneme_words[0] ?? "";
    const firstWordScore = firstWord ? ratio(noSpaceText, firstWord) : 0;
    return Math.max(prefix, firstWordScore);
  }

  private _continuationBonuses(
    hint: [number, number] | null,
  ): Map<string, number> {
    const bonuses = new Map<string, number>();
    if (!hint) return bonuses;

    const [hSurah, hAyah] = hint;
    const nv = this._byRef.get(`${hSurah}:${hAyah + 1}`);
    if (nv) {
      bonuses.set(`${hSurah}:${hAyah + 1}`, 0.22);
      if (this._byRef.has(`${hSurah}:${hAyah + 2}`))
        bonuses.set(`${hSurah}:${hAyah + 2}`, 0.12);
      if (this._byRef.has(`${hSurah}:${hAyah + 3}`))
        bonuses.set(`${hSurah}:${hAyah + 3}`, 0.06);
    } else {
      const nextVerses = this._bySurah.get(hSurah + 1) ?? [];
      const bonusValues = [0.22, 0.12, 0.06];
      for (let i = 0; i < Math.min(nextVerses.length, 3); i++) {
        bonuses.set(
          `${nextVerses[i].surah}:${nextVerses[i].ayah}`,
          bonusValues[i],
        );
      }
    }
    return bonuses;
  }

  private static _suffixPrefixScore(text: string, verseText: string): number {
    const wordsT = text.split(" ");
    const wordsV = verseText.split(" ");
    if (wordsT.length < 2 || wordsV.length < 2) return 0.0;

    let best = 0.0;
    const maxTrim = Math.min(Math.floor(wordsT.length / 2), 4);
    for (let trim = 1; trim <= maxTrim; trim++) {
      const suffix = wordsT.slice(trim).join(" ");
      const n = wordsT.length - trim;
      const prefix = wordsV.slice(0, Math.min(n, wordsV.length)).join(" ");
      best = Math.max(best, ratio(suffix, prefix));
    }
    return best;
  }
}

File: shared/quran_db.py
import array
import json
import math
from collections import defaultdict
from pathlib import Path
from Levenshtein import ratio
from shared.normalizer import normalize_arabic


def partial_ratio(short: str, long: str) -> float:
    """Levenshtein ratio of *short* against its best-matching window in *long*.

    Useful for detecting when a short transcription is a fragment of a longer
    verse that was already emitted.
    """
    if not short or not long:
        return 0.0
    if len(short) > len(long):
        short, long = long, short
    window = len(short)
    best = 0.0
    for i in range(max(1, len(long) - window + 1)):
        r = ratio(short, long[i:i + window])
        if r > best:
            best = r
            if best == 1.0:
                break
    return best

# Resolve to project root / data / quran.json
DATA_PATH = Path(__file__).parent.parent / "data" / "quran.json"


_BSM_CLEAN = normalize_arabic("بسم الله الرحمن الرحيم")
_FRAGMENT_MIN_WORDS = 4
_FRAGMENT_PARTIAL_WEIGHT = 0.75


class QuranDB:
    def __init__(self, path: Path = DATA_PATH):
        with open(path) as f:
            self.verses = json.load(f)
        self._by_ref = {}
        self._by_surah = {}
        # Strip BOM from text_clean (quran.json verse 1:1 has a stray BOM)
        for v in self.verses:
            v["text_clean"] = v["text_clean"].lstrip('\ufeff')
        for v in self.verses:
            self._by_ref[(v["surah"], v["ayah"])] = v
            self._by_surah.setdefault(v["surah"], []).append(v)
            # Pre-compute bismillah-stripped text for verse 1 of each surah
            # (Al-Fatiha 1:1 IS the bismillah, At-Tawbah 9 has none)
            if (
                v["ayah"] == 1
                and v["surah"] not in (1, 9)
                and v["text_clean"].startswith(_BSM_CLEAN)
            ):
                stripped = v["text_clean"][len(_BSM_CLEAN):].strip()
                v["text_clean_no_bsm"] = stripped if stripped else None
            else:
                v["text_clean_no_bsm"] = None

        # Build reverse-lookup and trigram index for fast candidate retrieval
        self._ref_to_idx: dict[tuple, int] = {}
        for i, v in enumerate(self.verses):
            self._ref_to_idx[(v["surah"], v["ayah"])] = i
        self._build_trigram_index()

    @property
    def total_verses(self):
        return len(self.verses)

    @property
    def surah_count(self):
        return len(self._by_surah)

    def get_verse(self, surah: int, ayah: int):
        return self._by_ref.get((surah, ayah))

    def get_surah(self, surah: int):
        return self._by_surah.get(surah, [])

    def get_next_verse(self, surah: int, ayah: int) -> dict | None:
        """Return the next verse after surah:ayah, or None if last verse."""
        verses = self._by_surah.get(surah, [])
        for i, v in enumerate(verses):
            if v["ayah"] == ayah:
                if i + 1 < len(verses):
                    return verses[i + 1]
                next_surah = self._by_surah.get(surah + 1, [])
                return next_surah[0] if next_surah else None
        return None

    def search(self, text: str, top_k: int = 5) -> list[dict]:
        text = normalize_arabic(text)
        scored = []
        for v in self.verses:
            score = ratio(text, v["text_clean"])
            scored.append({**v, "score": score, "text": v["text_uthmani"]})
        scored.sort(key=lambda x: x["score"], reverse=True)
        return scored[:top_k]

    def _continuation_bonuses(
        self, hint: tuple[int, int] | None
    ) -> dict[tuple[int, int], float]:
        """Build a map of (surah, ayah) → score bonus for expected next verses."""
        if not hint:
            return {}
        h_surah, h_ayah = hint
        bonuses: dict[tuple[int, int], float] = {}
        nv = self._by_ref.get((h_surah, h_ayah + 1))
        if nv:
            bonuses[(h_surah, h_ayah + 1)] = 0.22
            if self._by_ref.get((h_surah, h_ayah + 2)):
                bonuses[(h_surah, h_ayah + 2)] = 0.12
            if self._by_ref.get((h_surah, h_ayah + 3)):
                bonuses[(h_surah, h_ayah + 3)] = 0.06
        else:
            # Last ayah in surah — bonus carries to first ayah(s) of next surah
            next_verses = self._by_surah.get(h_surah + 1, [])
            for i, nv in enumerate(next_verses[:3]):
                bonus = [0.22, 0.12, 0.06][i]
                bonuses[(nv["surah"], nv["ayah"])] = bonus
        return bonuses

    @staticmethod
    def _char_trigrams(text: str) -> set[str]:
        """Extract character trigrams from text."""
        return {text[i:i + 3] for i in range(len(text) - 2)} if len(text) >= 3 else set()

    def _build_trigram_index(self):
        """Build char-trigram inverted index over all verses for fast retrieval."""
        posting: dict[str, set[int]] = defaultdict(set)
        N = len(self.verses)
        for idx, v in enumerate(self.verses):
            trigrams = self._char_trigrams(v["text_clean"])
            if v.get("text_clean_no_bsm"):
                trigrams |= self._char_trigrams(v["text_clean_no_bsm"])
            for tri in trigrams:
                posting[tri].add(idx)
        # Convert to sorted array.array('H') for compactness (~3MB)
        self._trigram_index: dict[str, array.array] = {}
        self._idf: dict[str, float] = {}
        for tri, indices in posting.items():
            self._trigram_index[tri] = array.array('H', sorted(indices))
            self._idf[tri] = math.log(N / len(indices))

    def _trigram_candidates(self, text: str, top_k: int = 50) -> list[int]:
        """Return top_k verse indices by IDF-weighted trigram overlap with text."""
        trigrams = self._char_trigrams(text)
        if not trigrams:
            return []
        scores: dict[int, float] = {}
        for tri in trigrams:
            if tri not in self._trigram_index:
                continue
            w = self._idf[tri]
            for idx in self._trigram_index[tri]:
                scores[idx] = scores.get(idx, 0.0) + w
        ranked = sorted(scores, key=scores.__getitem__, reverse=True)
        return ranked[:top_k]

    @staticmethod
    def _suffix_prefix_score(text: str, verse_text: str) -> float:
        """Best Levenshtein ratio from matching suffixes of *text* against
        equal-length prefixes of *verse_text*.

        After a window reset the transcription often starts with residual
        words from the *previous* verse followed by the start of the *next*
        verse. This method finds the best alignment by sliding the split
        point through the transcription.
        """
        words_t = text.split()
        words_v = verse_text.split()
        if len(words_t) < 2 or len(words_v) < 2:
            return 0.0
        best = 0.0
        max_trim = min(len(words_t) // 2, 4)
        for trim in range(1, max_trim + 1):
            suffix = " ".join(words_t[trim:])
            n = len(words_t) - trim
            prefix = " ".join(words_v[:min(n, len(words_v))])
            best = max(best, ratio(suffix, prefix))
        return best

    @staticmethod
    def _fragment_score(text: str, verse_text: str, full_ratio: float) -> float:
        """Score a partial transcript against a longer verse.

        Streaming windows often contain only a fragment of a long ayah. A
        full-string Levenshtein ratio suppresses those matches too heavily, so
        blend in partial matching when the query is long enough to be specific.
        Short candidate verses are penalized to avoid spuriously matching a
        tiny verse that happens to occur inside a longer transcript.
        """
        query_words = text.split()
        verse_words = verse_text.split()
        if len(query_words) < _FRAGMENT_MIN_WORDS or len(verse_words) < 2:
            return full_ratio

        fragment = partial_ratio(text, verse_text)
        if fragment <= full_ratio:
            return full_ratio

        shorter_penalty = min(1.0, len(verse_words) / max(len(query_words), 1))
        blended = (
            (1.0 - _FRAGMENT_PARTIAL_WEIGHT) * full_ratio
            + _FRAGMENT_PARTIAL_WEIGHT * fragment * shorter_penalty
        )
        return max(full_ratio, blended)

    def match_verse(
        self,
        text: str,
        threshold: float = 0.3,
        max_span: int = 3,
        hint: tuple[int, int] | None = None,
        return_top_k: int = 0,
    ) -> dict | None:
        """Find the best matching verse or consecutive verse span.

        Two-pass: first find top single-verse candidates (fast), then try
        multi-ayah spans only around those candidates.

        If *hint* is provided as (surah, ayah) of the last matched verse,
        the expected next verses receive a score bonus so sequential
        recitation is favoured over re-inferring from scratch.

        If *return_top_k* > 0, the returned dict includes a ``"runners_up"``
        list with the next-best candidates (each with raw_score and bonus).
        """
        text = normalize_arabic(text)
        if not text.strip():
            return None

        bonuses = self._continuation_bonuses(hint)

        # Pass 1: trigram retrieval → Levenshtein rerank on small candidate set
        candidate_idxs = set(self._trigram_candidates(text, top_k=50))

        # Always inject continuation bonus candidates
        for ref in bonuses:
            idx = self._ref_to_idx.get(ref)
            if idx is not None:
                candidate_idxs.add(idx)

        # Fallback: if trigram retrieval returns too few, full scan
        if len(candidate_idxs) < 20:
            candidate_idxs = set(range(len(self.verses)))

        scored = []
        for idx in candidate_idxs:
            v = self.verses[idx]
            raw = ratio(text, v["text_clean"])
            raw = self._fragment_score(text, v["text_clean"], raw)
            if v["text_clean_no_bsm"]:
                stripped_raw = ratio(text, v["text_clean_no_bsm"])
                stripped_raw = self._fragment_score(
                    text,
                    v["text_clean_no_bsm"],
                    stripped_raw,
                )
                raw = max(raw, stripped_raw)
            bonus = bonuses.get((v["surah"], v["ayah"]), 0.0)
            if bonus > 0:
                sp = self._suffix_prefix_score(text, v["text_clean"])
                raw = max(raw, sp)
            scored.append((v, raw, bonus, min(raw + bonus, 1.0)))
        scored.sort(key=lambda x: x[3], reverse=True)

        if not scored:
            return None

        best_v, best_raw, best_bonus, best_score = scored[0]
        best = {**best_v, "score": best_score, "raw_score": best_raw, "bonus": best_bonus}

        # Collect single-verse runners-up before span pass
        top_singles = [
            {
                "surah": v["surah"],
                "ayah": v["ayah"],
                "raw_score": round(raw, 3),
                "bonus": round(bon, 3),
                "score": round(total, 3),
                "text_clean": v["text_clean"][:60],
            }
            for v, raw, bon, total in scored[:max(return_top_k, 5)]
        ]

        # Pass 2: try multi-ayah spans around all candidate surahs
        # (candidate set is already small from trigram retrieval)
        seen_surahs = set()
        for v, _raw, _bon, _total in scored:
            s = v["surah"]
            if s in seen_surahs:
                continue
            seen_surahs.add(s)
            verses = self._by_surah[s]
            for i, sv in enumerate(verses):
                for span in range(2, max_span + 1):
                    if i + span > len(verses):
                        break
                    chunk = verses[i:i + span]
                    # Use no-bismillah text for the first verse in a span
                    first_text = (
                        chunk[0]["text_clean_no_bsm"] or chunk[0]["text_clean"]
                    )
                    combined = " ".join(
                        [first_text] + [c["text_clean"] for c in chunk[1:]]
                    )
                    raw = ratio(text, combined)
                    bonus = bonuses.get((chunk[0]["surah"], chunk[0]["ayah"]), 0.0)
                    score = min(raw + bonus, 1.0)
                    if score > best_score:
                        best_score = score
                        best = {
                            "surah": s,
                            "ayah": chunk[0]["ayah"],
                            "ayah_end": chunk[-1]["ayah"],
                            "text": " ".join(c["text_uthmani"] for c in chunk),
                            "text_clean": combined,
                            "score": score,
                            "raw_score": raw,
                            "bonus": bonus,
                        }

        if best_score >= threshold:
            if return_top_k > 0:
                best["runners_up"] = top_singles[:return_top_k]
            return best
        return None
